myia/ir/manager.py

```diff
@@ -0,0 +1,845 @@
+    """Represents a graph's immediate parent."""
+
+    def __init__(self, graph):
+        """Initialize the ParentProxy."""
+        self.graph = graph
+
+    def __hash__(self):
+        return hash(self.graph)
+
+    def __eq__(self, other):
+        return isinstance(other, ParentProxy) and other.graph is self.graph
+
+
+class PerGraphStatistic(dict):
+    """Represents a statistic that maps each graph to some information."""
+
+    constructor = dict
+    include_graph_none = False
+
+    def __init__(self, manager):
+        """Initialize a PerGraphStatistic."""
+        super().__init__()
+        self.manager = manager
+        evts = self.manager.events
+        evts.add_node.register(self._on_add_node)
+        evts.drop_node.register(self._on_drop_node)
+        evts.add_graph.register(self._on_add_graph)
+        evts.drop_graph.register(self._on_drop_graph)
+        evts.add_edge.register(self._on_add_edge)
+        evts.drop_edge.register(self._on_drop_edge)
+        if self.include_graph_none:
+            self[None] = self.constructor()
+
+    def reset(self):
+        """Reset this graph's information."""
+        return self.clear()
+
+    def _on_add_graph(self, event, graph):
+        self[graph] = self.constructor()
+
+    def _on_drop_graph(self, event, graph):
+        del self[graph]
+
+    def _on_add_node(self, event, node):
+        pass
+
+    def _on_drop_node(self, event, node):
+        pass
+
+    def _on_add_edge(self, event, node, key, value):
+        pass
+
+    def _on_drop_edge(self, event, node, key, value):
+        pass
+
+
+class NodesStatistic(PerGraphStatistic):
+    """Implements `GraphManager.nodes`."""
+
+    constructor = set
+    include_graph_none = True
+
+    def _on_add_node(self, event, node):
+        self[node.graph].add(node)
+
+    def _on_drop_node(self, event, node):
+        self[node.graph].remove(node)
+
+
+class CounterStatistic(PerGraphStatistic):
+    """Represents a statistic that maps each graph to a set of counters."""
+
+    constructor = dict
+
+    def inc(self, graph, key, qty=1):
+        """Increment the count for self[graph][key] by qty."""
+        d = self[graph]
+        if key not in d:
+            d[key] = qty
+            return True
+        else:
+            d[key] += qty
+            return False
+
+    def dec(self, graph, key, qty=1):
+        """Decrement the count for self[graph][key] by qty.
+
+        The key is deleted if the count falls to zero.
+        """
+        d = self[graph]
+        if d[key] == qty:
+            del d[key]
+            return True
+        else:
+            d[key] -= qty
+            return False
+
+    def mod(self, graph, key, qty):
+        """Change the count for self[graph][key] by qty."""
+        if qty > 0:
+            return self.inc(graph, key, qty)
+        elif qty < 0:
+            return self.dec(graph, key, -qty)
+        else:
+            raise ValueError('qty cannot be 0')  # pragma: no cover
```

Comment:
If there is a legitimate case where qty is 0, yes, but right now there isn't. In most cases qty is 1 or -1. There's also code that updates using counts from another dictionary, but since keys should be removed when count falls to 0, these counts should never be 0. Anyway, I'd like to be notified when this happens, because it means there's probably a bug.